### PR TITLE
Support using `cargo make` as library

### LINKED ISSUE #1110: `cargo-make` as library: task exection in asynchronous persistent queues
### Feature Description
`cargo-make` is great. I've been using it since at least 2018. I have ideas that are too far outside the scope of `cargo-make` to be contributed back to `master`.

### Describe The Solution You'd Like
Can we expose `cli::descriptor::load` https://github.com/sagiegurari/cargo-make/blob/153fa22/src/lib/descriptor/mod.rs#L581-L586 ? - A bunch of other modules also? - Your `src/lib/cli.rs` uses a lot of private and `pub(crate)` `mod`ules.

### Code Sample
https://github.com/offscale/cargo-make/tree/publicise

---

I have ideas about making tasks asynchronous whilst retaining parallelizability by using a real database or Redis (valkey) to continue/resume/replay. But need to either fork your repo (sad) or get you to expose internals in order to experiment here.

Thanks for your consideration

### PR DESCRIPTION
**Pull request by SamuelMarks**
Related to and closes #1110

### CODE REVIEW

File: src/lib/error.rs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+{
+    fn from_residual(residual: Result<std::convert::Infallible, std::io::Error>) -> Self {
+        if let Err(e) = residual {
+            SuccessOrCargoMakeError::Err(e.into())
+        } else {
+            SuccessOrCargoMakeError::Ok(std::process::ExitCode::SUCCESS)
+        }
+    }
+}
+*/
```
Comment (sagiegurari) [inline src/lib/error.rs:194]: what is this todo about?

File: src/lib/types.rs [hunk 2]
```diff
@@ -2451,3 +2452,17 @@ pub struct CommandSpec {
...
+        Either::Right(rhs) => Err(rhs),
+    }
+}
+
+pub fn result_to_either<V, E>(result: Result<V, E>) -> Either<V, E> {
+    match result {
+        Ok(lhs) => Either::Left(lhs),
+        Err(rhs) => Either::Right(rhs),
+    }
+}
```
Comment (sagiegurari) [inline src/lib/types.rs:2468]: who is using this?

**Comment by sagiegurari** [comment]
@SamuelMarks amazing job thanks a lot. its actually something i wanted to do for a long time but didn't have the capacity so its great. 
got few questions and also i see a test error. would love if you can take a look.

File: src/lib/types.rs (hunk 2 above)
Comment (SamuelMarks) [reply src/lib/types.rs:2468]: Oh that's leftover from when I was using `Either`; let me clean that up

File: src/lib/error.rs (hunk 1 above)
Comment (SamuelMarks) [reply src/lib/error.rs:194]: If you're ok with enabling `#![feature(try_trait_v2)]` then I will put that in, uncomment this, and use `?` in your `main` entrypoints.

**Comment by SamuelMarks** [comment]
@sagiegurari Great to hear you're open to this PR!

Let me take a look at the failing test and see if I can't resolve.

File: src/lib/error.rs (hunk 1 above)
Comment (sagiegurari) [reply src/lib/error.rs:194]: not familiar with it but i want to be backward compatible as much as possible including in rust versions. so lets drop it for now or if we can wrap it with some config that checks if feature is enabled?

Comment (SamuelMarks) [reply src/lib/error.rs:194]: Ok it's nightly only still it seems https://github.com/metalbear-co/mirrord/issues/1568 https://github.com/rust-lang/rust/issues/84277 so will remove the commented out lines. This PR is ready to merge.

**Comment by sagiegurari** [comment]
@SamuelMarks your fixes solve the CI issue but i also merged #1109 which caused some merge conflicts. any chance you can resolve it so i'll push this? thanks

**Comment by SamuelMarks** [comment]
@sagiegurari Resolved that merge conflict, still some weirdness going with this logic:
https://github.com/sagiegurari/cargo-make/blob/0119754/src/lib/execution_plan.rs#L491-L498

Once resolved let's try and merge this quickly before more PRs come through! - As this PR changes the whole codebase, adding error handling &etc.

**Comment by sagiegurari** [comment]
thanks. there are no PRs at the moment so i wouldn't worry about it.
as for the legacy task, i'm planning to remove it soon i hope.

**Comment by SamuelMarks** [comment]
Cool cool; no I mean 18 tests are failing because of that logic I linked. What do you want that logic to be? - Silent ignore? - Ignore but give a warning message? - Something else?

**Comment by sagiegurari** [comment]
I'll review the updated code tomorrow. if it makes a fuzz, drop the legacy migration task from the codebase as i want to delete it anyway.

**Comment by sagiegurari** [comment]
@SamuelMarks merging. thanks a lot

**Comment by sagiegurari** [comment]
@SamuelMarks while merging it to master (and planning to release it), i saw that there is an unsafe block and its crashing in mac OS in the ci cd flow.
please take a look

https://github.com/sagiegurari/cargo-make/actions/runs/9970773762

**Pull request merged**